File: libint-sys/shim/src/atom.cc
#include "libint-sys/shim/include/atom.h"
#include <sstream>

namespace libint2 {
std::unique_ptr<Atom> atom(const int atomic_number, const double x,
                           const double y, const double z) {
  return std::make_unique<Atom>(Atom{atomic_number, x, y, z});
}

int atomic_number(const Atom &atom) { return atom.atomic_number; }

double x(const Atom &atom) { return atom.x; }

double y(const Atom &atom) { return atom.y; }

double z(const Atom &atom) { return atom.z; }

void scale(Atom& atom, double factor) {
  atom.x *= factor;
  atom.y *= factor;
  atom.z *= factor;
}

// std::unique_ptr<std::vector<Atom>> read_dotxyz(const std::string& is) {
//   std::istringstream stream(is);
//   return std::make_unique<std::vector<Atom>>(read_dotxyz(stream));
// }
} // namespace libint2

File: libint-sys/src/bridge/atom.rs
use cxx::UniquePtr;

#[cxx::bridge(namespace = "libint2")]
pub mod ffi {
    unsafe extern "C++" {
        include!("libint-sys/shim/include/atom.h");

        type Atom;

        #[must_use]
        fn atom(atomic_number: i32, x: f64, y: f64, z: f64) -> UniquePtr<Atom>;

        // getters
        #[must_use]
        fn atomic_number(atom: &Atom) -> i32;
        #[must_use]
        fn x(atom: &Atom) -> f64;
        #[must_use]
        fn y(atom: &Atom) -> f64;
        #[must_use]
        fn z(atom: &Atom) -> f64;
    }
}

#[cfg(test)]
mod tests {
    use super::ffi;

    #[test]
    #[allow(clippy::float_cmp)]
    fn atom() {
        let atom = ffi::atom(1, 0., 1., 2.);
        assert_eq!(ffi::atomic_number(&atom), 1);
        assert_eq!(ffi::x(&atom), 0.);
        assert_eq!(ffi::y(&atom), 1.);
        assert_eq!(ffi::z(&atom), 2.);
    }
}
